spark2/src/main/java/org/apache/iceberg/spark/source/StreamingReader.java

```diff
@@ -0,0 +1,288 @@
+    MicroBatch lastBatch = batches.isEmpty() ? null : batches.get(batches.size() - 1);
+
+    if (lastBatch == null) {
+      return start;
+    } else {
+      boolean isStarting = lastBatch.snapshotId() == start.snapshotId() && start.isStartingSnapshotId();
+      return new StreamingOffset(lastBatch.snapshotId(), lastBatch.endFileIndex(), isStarting,
+          lastBatch.lastIndexOfSnapshot());
+    }
+  }
+
+  @VisibleForTesting
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
+  List<MicroBatch> getChangesWithRateLimit(long snapshotId, int index, boolean isStarting,
+                                           boolean isLastIndexOfSnapshot, long maxSize) {
+    List<MicroBatch> batches = Lists.newArrayList();
+    long currentLeftSize = maxSize;
+    MicroBatch lastBatch = null;
+
+    if (!isLastIndexOfSnapshot && (isStarting || isValidSnapshot(table.snapshot(snapshotId)))) {
+      MicroBatch batch = MicroBatches.from(table.snapshot(snapshotId), table.io())
+          .caseSensitive(super.caseSensitive())
+          .specsById(table.specs())
+          .generate(index, currentLeftSize, isStarting);
+
+      batches.add(batch);
+      currentLeftSize -= batch.sizeInBytes();
+      lastBatch = batch;
+    }
+
+    // Current snapshot can already satisfy the size needs.
+    // Or DataFile size cannot satisfy the current needs of max size. For example: the request left size is 100, but the
+    // file size is 200, then file cannot add to this batch.
+    if (currentLeftSize <= 0L || (lastBatch != null && !lastBatch.lastIndexOfSnapshot())) {
+      return batches;
+    }
+
+    long currentSnapshotId = table.currentSnapshot().snapshotId();
+    if (currentSnapshotId == snapshotId) {
+      // the snapshot of current offset is already the latest snapshot of this table.
+      return batches;
+    }
+
+    ImmutableList<Long> snapshotIds = ImmutableList.<Long>builder()
+        .addAll(SnapshotUtil.snapshotIdsBetween(table, snapshotId, currentSnapshotId))
+        .build()
+        .reverse();
+
+    for (Long id : snapshotIds) {
+      Snapshot snapshot = table.snapshot(id);
+      if (!isValidSnapshot(snapshot)) {
+        continue;
+      }
+
+      int startIndex = lastBatch == null || lastBatch.lastIndexOfSnapshot() ? 0 : lastBatch.endFileIndex();
+      MicroBatch batch = MicroBatches.from(table.snapshot(id), table.io())
+          .caseSensitive(super.caseSensitive())
+          .specsById(table.specs())
+          .generate(startIndex, currentLeftSize, false);
+
+      batches.add(batch);
+      currentLeftSize -= batch.sizeInBytes();
+
+      // If the current request size is already satisfied, or none of the DataFile size can satisfy the current left
+      // size, break the current loop.
+      if (currentLeftSize <= 0L || !batch.lastIndexOfSnapshot()) {
+        break;
+      }
+
+      lastBatch = batch;
+    }
+
+    return batches;
+  }
+
+  private static boolean isValidSnapshot(Snapshot snapshot) {
```

Comment:
I think this method does too much. Looks like it is used to skip non-append commits, but also to fail the read when an overwrite commit is encountered. The exception is surprising behavior given the method contract.

I think this should be divided into a validation that throws an exception if there is an overwrite, and a different method to skip rewrite commits.